server/jdkhttp-server/src/main/scala/sttp/tapir/server/jdkhttp/internal/ParsedMultiPart.scala

```diff
@@ -44,48 +49,125 @@ case class ParsedMultiPart(headers: Map[String, Seq[String]], body: Array[Byte])
 }
 
 object ParsedMultiPart {
-  def empty: ParsedMultiPart = new ParsedMultiPart(Map.empty, Array.empty)
+  def empty: ParsedMultiPart = new ParsedMultiPart(
+    Map.empty,
+    new ByteArrayInputStream(Array.empty)
+  )
 
   sealed trait ParseState
   case object Default extends ParseState
   case object AfterBoundary extends ParseState
+  case object AfterParsedBody extends ParseState
   case object AfterHeaderSpace extends ParseState
 
   private case class ParseData(
       currentPart: ParsedMultiPart,
       completedParts: List[ParsedMultiPart],
       parseState: ParseState
   ) {
+    def done(): ParseData = this
     def changeState(state: ParseState): ParseData = this.copy(parseState = state)
     def addHeader(header: String): ParseData = this.copy(currentPart = currentPart.addHeader(header))
-    def addBody(body: Array[Byte]): ParseData = this.copy(currentPart = currentPart.copy(body = currentPart.body ++ body))
-    def completePart(): ParseData = this.currentPart.getName match {
+    def completePart(body: InputStream): ParseData = this.currentPart.getName match {
       case Some(_) =>
         this.copy(
-          completedParts = completedParts :+ currentPart,
+          completedParts = completedParts :+ currentPart.copy(body = body),
           currentPart = empty,
-          parseState = AfterBoundary
+          parseState = AfterParsedBody
         )
-      case None => changeState(AfterBoundary)
+      case None => changeState(AfterParsedBody)
     }
   }
 
-  def parseMultipartBody(httpExchange: HttpExchange, boundary: String): Seq[ParsedMultiPart] = {
-    val reader = new BufferedReader(new InputStreamReader(httpExchange.getRequestBody))
-    val initialParseState: ParseData = ParseData(empty, List.empty, Default)
+  private def toTruncatedInputStream(streamAndFile: PartStream, stopAt: Int): InputStream = streamAndFile match {
+    case ByteStream(stream) =>
+      val bytes = stream.toByteArray
+      new ByteArrayInputStream(bytes.slice(0, stopAt))
+    case FileStream(f, fileOutputStream) =>
+      fileOutputStream.getChannel.truncate(stopAt.toLong)
+      fileOutputStream.close()
+      new FileInputStream(f)
+  }
+
+  private def readUntilNewline(inputStream: InputStream): Array[Byte] =
```

Comment:
I'm just starting to read through the code, but doesn't this mean that if we get a line which is over 52MB, we'll read it into memory anyway? What if there are no newlines in the part data for a long time?

Reply:
The `readUntilNewline` and `readStringUntilNewline` functions are only called in places where there is expected to be a newline in a few bytes if the body is correctly formed.
When parsing headers, when 
We could have a problem with a malformed body i suppose, I'll see if i can figure out a way to either throw an error in those cases or rework the parsing somehow :+1:

Reply:
Sounds good, maybe simply a hard limit - something reasonable that would cover the max boundary / header size? Or in case of blank line separators, anything above 1 is already an error

Reply:
Max boundary size is 70 characters (+ a few for delimiters) according to [the rfc](https://datatracker.ietf.org/doc/html/rfc1521#page-30). 
The headers doesn't really have a specced max-size as far as i could tell, but i saw a few defaulting to 4kb and 8kb max sizes for normal http headers, so i don't think something like that for a max size sounds like a terrible idea.

Reply:
Yeah 8KB sounds very reasonable